lib/spack/spack/variant.py

```diff
@@ -22,17 +22,552 @@
+            allowed = self.values + (self.default,)
+            self.single_value_validator = lambda x: x in allowed
+
+        self.exclusive = exclusive
+        self.group_validator = validator
+
+    def validate_or_raise(self, vspec, pkg=None):
+        """Validate a variant spec against this package variant. Raises an
+        exception if any error is found.
+
+        :param VariantSpec vspec: instance to be validated
+        :param Package pkg: the package that required the validation,
+            if available
+
+        :raises InconsistentValidationError: if vspec.name doesn't
+            match self.name
+        :raises MultipleValuesInExclusiveVariantError: if vspec has
+            multiple values but self.exclusive ==True
+        :raises InvalidVariantValueError: if vspec.value contains
+            invalid values
+        """
+        # Check the name of the variant
+        if self.name != vspec.name:
+            raise InconsistentValidationError(vspec, self)
+
+        # Check the values of the variant spec
+        value = vspec.value
+        if isinstance(vspec.value, (bool, str)):
+            value = (vspec.value,)
+
+        # If the value is exclusive there must be at most one
+        if self.exclusive and len(value) != 1:
+            raise MultipleValuesInExclusiveVariantError(vspec, pkg)
+
+        # Check and record the values that are not allowed
+        not_allowed_values = [
+            x for x in value if not self.single_value_validator(x)
+        ]
+        if not_allowed_values:
+            raise InvalidVariantValueError(self, not_allowed_values, pkg)
+
+        # Validate the group of values if needed
+        if self.group_validator is not None:
+            self.group_validator(value)
+
+    @property
+    def allowed_values(self):
+        """Returns a string representation of the allowed values for
+        printing purposes
+
+        :return: representation of the allowed values
+        :rtype: str
+        """
+        # Join an explicit set of allowed values
+        if self.values is not None:
+            v = tuple(str(x) for x in self.values)
+            return ', '.join(v)
+        # In case we were given a single-value validator
+        # print the docstring
+        docstring = inspect.getdoc(self.single_value_validator)
+        v = docstring if docstring else ''
+        return v
+
+    def make_default(self):
+        return self.make_variant(self.default)
+
+    def make_variant(self, value):
+        return self.variant_cls(self.name, value)
+
+    @property
+    def variant_cls(self):
+        if self.exclusive is False:
+            return MultiValuedVariant
+        elif self.values == (True, False):
+            return BoolValuedVariant
+        return SingleValuedVariant
+
+
+@lang.key_ordering
+class MultiValuedVariant(object):
+
+    @staticmethod
+    def from_node_dict(name, value):
+        if isinstance(value, list):
+            value = ','.join(value)
```

Comment:
Here I have just to check:
```python
if isinstance(value, list):
    value = ','.join(value)
    ...
```
If I move it to `_value_setter` I think I'll need to check instead:
```python
if isinstance(value, collections.Iterable) and not isinstance(value, str):
    t = value
else:
    t = re.split(...)
...
```
I slightly prefer this version, but if you have strong opinion on this issue I'll change it.